src/adc.rs

```diff
@@ -0,0 +1,562 @@
+//! # Work in Progress
+//! API for the ADC (Analog to Digital Converter)
+//!
+//! Note that the more specific your hardware selection is
+//! (e.g. stm32f303**xc** instead of just stm32f303)
+//! the more functionality is accessible
+//! (in this case: ADC3 and ADC4 additionaly to ADC1 and ADC2).
+//!
+//! # Examples
+//! For a simple, working example check `adc.rs` in the examples folder.
+use cortex_m::asm;
+use embedded_hal::adc::{Channel, OneShot};
+#[cfg(feature = "stm32f303")]
+use stm32f3::stm32f303::adc1_2::ccr::CKMODE_A;
+
+use crate::rcc::{Clocks, AHB};
+
+#[cfg(feature = "stm32f303")]
+const MAX_ADVREGEN_STARTUP_US: u32 = 10;
+
+#[cfg(feature = "stm32f303")]
+use crate::gpio::{gpioa, gpiob, gpioc, Analog};
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+use crate::gpio::{gpiod, gpioe, gpiof};
+
+#[cfg(feature = "stm32f303")]
+use crate::stm32::{ADC1, ADC1_2, ADC2};
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+use crate::stm32::{ADC3, ADC3_4, ADC4};
+
+/// ADC configuration
+///
+/// TODO: Remove `pub` from the register block once all functionalities are implemented.
+/// Leave it here until then as it allows easy access to the registers.
+pub struct Adc<ADC> {
+    pub rb: ADC,
+    clocks: Clocks,
+    ckmode: CKMODE,
+    operation_mode: Option<OperationMode>,
+}
+
+/// ADC sampling time
+///
+/// Each channel can be sampled with a different sample time.
+/// For Sampletime T_n the total conversion time (in ADC clock cycles) is
+/// 12.5 + (n + .5)
+///
+/// TODO: there are boundaries on how this can be set depending on the hardware.
+/// Check them and implement a sample time setting mechanism.
+pub enum SampleTime {
+    T_1,
+    T_2,
+    T_4,
+    T_7,
+    T_19,
+    T_61,
+    T_181,
+    T_601,
+}
+
+impl Default for SampleTime {
+    fn default() -> Self {
+        SampleTime::T_19
```

Comment:
The RM stated:
> For each channel, SMP[2:0] bits must be programmed to respect a minimum sampling time as specified in the ADC charateristics section of the datasheets.

So to avoid any problems there and to have a stable measurement, I decided to set it to T_19. Weather this is a sensible choice is up to debate, but I don't agree that using T_1 is a better one, as this is a *default* and not a *reset* value we're setting.

Reply:
I agree that we should deviate from the device's defaults if we can find safer ones. However, in this case we should:

a) Add a comment explaining why we choose this default
b) Make sure that T_19 is actually sensible

For b), the datasheet of the discovery says minimum sampling time is `1.5 / (ADC clock frequency)`. Which means it needs at minimum 1.5 ADC clock cycles for each sample, right? Which means at least for the discovery `T_1` would actually be the sensible default.

Now, to do this right we'd actually have to go through all stm32f303 datasheets and check their ADC characteristics. If we don't want to do that, we should either use the device default (`T_1`), or use the highest value just to be sure (`T_601`). Personally I think using the device default will usually be fine. I would rather not slow the sampling down to it's maximum if it's probably not necessary. But I also don't like randomly picking a value from the middle without any rationale.